File: balanced-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
'''
               3
            /     \ 
           9       20
                  /   \
                 15    7
h(9) 1

        1
         \
          2
           \
            3
'''
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        def height(node: Optional[TreeNode]) -> int:
            if not node:
                return 0
            if abs(height(node.left)-height(node.right))<=1:
                return 1+max(height(node.left),height(node.right))
            return float('inf')
        if not root:
            return True
        return abs(height(root.left)-height(root.right))<=1

File: balanced-binary-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
// left depth - right depth cannot differ by more than one!
// nil - depth is zero

//  1 
// 2 3
//      4
//         5
// should be true
// 1 + 1 + 1 + 
// dep(4)
// 2 -- 1
// 3 -- 3
// 1 - 3 == -2
// 1
//  2
//    3
// 1: ld: 0, rd: 
// 2: 
// 3: 
func dep(root *TreeNode) (int, bool) {
    if root == nil { return 0, true }
    ld, lans := dep(root.Left)
    rd, rans := dep(root.Right)
    val := ld - rd
    val = max(val, -val)
    return 1 + max(ld, rd), val < 2 && lans && rans
}

func isBalanced(root *TreeNode) bool {
    _, ans := dep(root)
    return ans
}